python_modules/dagster/dagster/_core/definitions/asset_condition/parent_outdated.py

```diff
@@ -0,0 +1,79 @@
+    AncestorUnsynced or something of the sort.
+    """
+
+    @property
+    def description(self) -> str:
+        return ""
+
+    def evaluate(self, context: NewAssetConditionEvaluationContext) -> "AssetConditionResult":
+        slice_to_evaluate = (
+            # evaluate any candidate partitions that have parents which have been updated since
+            # the previous tick
+            context.candidate_slice & context.compute_parent_updated_since_previous_tick_slice()
+        ) | (
+            # also evaluate any candidate partitions which were not evaluated in the previous tick
+            context.candidate_slice - context.previous_candidate_subset
+        )
+
+        outdated_ancestors_by_asset_partition_key: Dict[AssetKeyPartitionKey, Set[AssetKey]] = (
+            defaultdict(set)
+        )
+
+        parent_keys_to_inspect = {
+            key
+            for key in context._asset_graph.get(context.asset_key).parent_keys
+            if not context._instance_queryer.have_ignorable_partition_mapping_for_outdated(
+                context.asset_key, key
+            )
+        }
+        for asset_partition_key in slice_to_evaluate.compute_asset_partition_keys():
+            # slice representing the single asset partition key we're evaluating
+            current_slice = context.asset_graph_view.get_asset_slice_from_asset_partition_keys(
+                context.asset_key, {asset_partition_key}
+            )
+            for parent_asset_key in parent_keys_to_inspect:
+                parent_slice = current_slice.compute_parent_slice(parent_asset_key)
+                for parent_asset_partition_key in parent_slice.compute_asset_partition_keys():
+                    outdated_ancestors_by_asset_partition_key[asset_partition_key].update(
+                        context._instance_queryer.get_outdated_ancestors(
+                            asset_partition=parent_asset_partition_key
+                        )
+                    )
+
+        return AssetConditionResult.create_from_previous_evaluation(
+            context,
+            new_evaluated_slice=slice_to_evaluate,
+            new_true_slice=context.asset_graph_view.get_asset_slice_from_asset_partition_keys(
+                context.asset_key,
+                {
+                    apk
+                    for apk, outdated_ancestors in outdated_ancestors_by_asset_partition_key.items()
+                    if outdated_ancestors
+                },
+            ),
+            new_metadata_by_asset_partition_key={
+                asset_partition: {
+                    f"waiting_on_ancestor_{i+1}": MetadataValue.asset(k)
+                    for i, k in enumerate(sorted(outdated_ancestors))
+                }
+                for asset_partition, outdated_ancestors in outdated_ancestors_by_asset_partition_key.items()
+                if outdated_ancestors
+            },
+        )
```

Comment:
@OwenKephart so when we chatted about this I was imagining the body of this logic being _much_ thinner. Basically pushing this down into the asset graph view/slice layer and having no direct contact with the instance queryer.

Reply:
```python
for asset_partition_key in slice_to_evaluate.compute_asset_partition_keys():
            # slice representing the single asset partition key we're evaluating
            current_slice = context.asset_graph_view.get_asset_slice_from_asset_partition_keys(
                context.asset_key, {asset_partition_key}
            )
            for parent_asset_key in parent_keys_to_inspect:
                parent_slice = current_slice.compute_parent_slice(parent_asset_key)
                for parent_asset_partition_key in parent_slice.compute_asset_partition_keys():
                    outdated_ancestors_by_asset_partition_key[asset_partition_key].update(
                        context._instance_queryer.get_outdated_ancestors(
                            asset_partition=parent_asset_partition_key
                        )
                    )
```

don't have everything hooked up yet but something like this feels a bit better. one thing to note is that we're not propagating metadata from the slice calls, meaning we don't know exactly which ancestor we're waiting on, we just know which parent is currently outdated in this version.

however, that aligns nicely with our desire to make things less global (i.e. the 'parent unsynced' rule would work this way)

Reply:
ok updated the pr and tests are passing.

definitely feels nicer -- some amount of business logic still needs to exist within the AssetCondition, as I think we'll eventually need to let users pick and choose exactly which parents they care about, and so I couldn't push all of this into just `slice_to_evaluate.compute_parent_outdated_slice()`. However, I think this works out fairly clean, and this pattern of:

- get parent slice
- get parent slice w/ property
- get child slice of that parent slice

can be reused for many different rules